stan-dev/stan
# Bugfix/issue 2178 fix nuts criterion

#### Submisison Checklist


#### Summary

Addresses #2178 by modifying the build_tree code to return proper tree boundaries and fixing the rho used for the right subtree.

#### Intended Effect

Restores exact validity of the NUTS sampler.  The NUTS tests were expanded to catch these issues.

#### How to Verify

Run unit tests in src/test/unit/mcmc/hmc/nuts.

Also, run the modified model I posted in the comments of #2178 to see that the two marginal variances and the correlation are both within the MCMC standard error, even with one million iterations.

#### Side Effects

Small changes to the sampler behavior which should decrease with the sampler step size.  May break the upstream integration test.

#### Documentation

None user facing.  Inline doc was updated.

#### Reviewer Suggestions

@syclik @bob-carpenter 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
